pkg/sentry/syscalls/linux/vfs2/io_uring.go

```diff
@@ -0,0 +1,139 @@
+	"gvisor.dev/gvisor/pkg/syserror"
+
+	"unsafe"
+)
+
+// Represent `struct io_uring` as defined in
+// https://elixir.bootlin.com/linux/v5.10.44/source/fs/io_uring.c#L107
+type IoUring struct {
+	Head uint32
+	_    [15]uint32 // Head and tail need to be "____cacheline_aligned_in_smp"
+	Tail uint32
+	_    [15]uint32 // Head and tail need to be "____cacheline_aligned_in_smp"
+}
+
+// Represent `struct io_rings` as defined in
+// https://elixir.bootlin.com/linux/v5.10.44/source/fs/io_uring.c#L119
+type IoRings struct {
+	Sq            IoUring
+	Cq            IoUring
+	SqRingMask    uint32
+	CqRingMask    uint32
+	SqRingEntries uint32
+	CqRingEntries uint32
+	SqDropped     uint32
+	SqFlags       uint32
+	CqFlags       uint32
+	CqOverflow    uint32
+	// TODO: Add cqes array?
+}
+
+// IoUringSetup implements Linux syscall io_uring_setup(2)
+func IoUringSetup(t *kernel.Task, args arch.SyscallArguments) (uintptr, *kernel.SyscallControl, error) {
+	entries := args[0].Uint()
+	params := args[1].Pointer()
+
+	paramsStruct, err := ioUringCreate(t, entries, params)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	file_io, err_io := iouringfs.NewIouringfsFile(t, t.Kernel().IoUringMount(), paramsStruct.SqEntries, paramsStruct.CqEntries)
+	if err_io != nil {
+		return 0, nil, err
+	}
+	fd_io, err_io := t.NewFDFromVFS2(0, file_io, kernel.FDFlags{
+		CloseOnExec: linux.O_CLOEXEC != 0,
+	})
+
+	// 2 * IORING_OFF_SQES is to have enought space on the tmpfs file. This
+	// should change once we implement `iouringfs`
+	// file, err := tmpfs.NewZeroFile(t, t.Credentials(), t.Kernel().ShmMount(),
+	// 	2*linux.IORING_OFF_SQES)
+	// if err != nil {
+	// 	return 0, nil, err
+	// }
+	//
+	// fd, err := t.NewFDFromVFS2(0, file, kernel.FDFlags{
+	// 	CloseOnExec: linux.O_CLOEXEC != 0,
+	// })
+	// if err != nil {
+	// 	return 0, nil, err
+	// }
+
+	return uintptr(fd_io), nil, nil
+
+}
+
+func ioUringCreate(t *kernel.Task, entries uint32, p hostarch.Addr) (*linux.IoUringParams, error) {
+	if entries == 0 {
+		return nil, syserror.EFAULT
+	}
+
+	var params linux.IoUringParams
+
+	_, err := params.CopyIn(t, p)
+	if err != nil {
+		return nil, err
+	}
+
+	if entries > linux.IORING_MAX_ENTRIES {
+		entries = linux.IORING_MAX_ENTRIES
+	}
+
+	params.SqEntries = entries
+	params.CqEntries = 2 * params.SqEntries
+
+	rings := IoRings{}
+
+	params.SqOff.Head = uint32(unsafe.Offsetof(rings.Sq.Head))
+	params.SqOff.Tail = uint32(unsafe.Offsetof(rings.Sq.Tail))
+	params.SqOff.RingMask = uint32(unsafe.Offsetof(rings.SqRingMask))
+	params.SqOff.RingEntries = uint32(unsafe.Offsetof(rings.SqRingEntries))
+	params.SqOff.Flags = uint32(unsafe.Offsetof(rings.SqFlags))
+	params.SqOff.Dropped = uint32(unsafe.Offsetof(rings.SqDropped))
+	params.SqOff.Array = 384 // TODO: Unconstify this
+
+	params.CqOff.Head = uint32(unsafe.Offsetof(rings.Cq) + unsafe.Offsetof(rings.Cq.Head)) // For some reason, directly using Offsetof(rigs.Cq.Head) does not work as excepted (or as in C)
```

Comment:
This should be fine here because the offsets are the used by the userland. So even if the memory layout is not the same as in C, userland should be able to communicate correctly with gvisor.

Reply:
> Are you arguing that since the sentry is essentially telling the user application the offsets here, so even if they don't match a c memory layout, things will be interpreted correctly?

Exactly. For example [here](https://github.com/shuveb/io_uring-by-example/blob/58d05c1137d672337cf671e03e73bf7cc1b96342/02_cat_uring/main.c#L174) we use the offset and to compute the real address of the tail and [later](https://github.com/shuveb/io_uring-by-example/blob/58d05c1137d672337cf671e03e73bf7cc1b96342/02_cat_uring/main.c#L327) we update the tail